### PR TITLE
[scheduler] Update task from push message

### PR DESCRIPTION
**Pull request by CaseyHillers**
Fixes https://github.com/flutter/flutter/issues/98288
Fixes https://github.com/flutter/flutter/issues/98287

### CODE REVIEW

File: app_dart/lib/src/model/appengine/task.dart [hunk 1]
```diff
@@ -262,27 +262,77 @@ class Task extends Model<int> {
     _status = value;
   }
 
-  /// Update [Task] related fields based on LUCI's [BuildPushMessage].
-  String updateFromBuildPushMessage(BuildPushMessage pushMessage) {
-    if (pushMessage.build?.status == Status.started) {
+  /// Update [Task] fields based on a LUCI [Build].
+  void updateFromBuild(Build build) {
+    final List<String>? tags = build.tags;
+    // Example tag: build_address:luci.flutter.prod/Linux Cocoon/271
```
Comment (keyonghan) [inline app_dart/lib/src/model/appengine/task.dart:268]: Is the tags the `Tags` on the build page: https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20framework_tests_widgets/3643/overview?

File: app_dart/lib/src/model/appengine/task.dart [hunk 2]
```diff
@@ -262,27 +262,77 @@ class Task extends Model<int> {
...
+      log.fine('Skipping message as build number is before the current task');
+      return;
+    }
+
+    if (buildNumberList == null) {
+      buildNumberList = '$buildNumber';
+    } else {
+      final List<String> deserializedBuildNumberList = buildNumberList!.split(',');
+      deserializedBuildNumberList.add('$buildNumber');
+      buildNumberList = deserializedBuildNumberList.join(',');
```
Comment (keyonghan) [inline app_dart/lib/src/model/appengine/task.dart:292]: `buildNumberList = '$buildNumberList, $buildNumber'`?

File: app_dart/lib/src/model/appengine/task.dart [hunk 3]
```diff
@@ -262,27 +262,77 @@ class Task extends Model<int> {
...
+    final String? buildAddress = tags?.firstWhere((String tag) => tag.contains('build_address'));
+    if (buildAddress == null) {
+      log.warning('Tags: $tags');
+      throw const BadRequestException('build_address does not contain build number');
+    }
+
+    final int currentBuildNumber = int.parse(buildAddress.split('/').last);
+    if (buildNumber == null || buildNumber! < currentBuildNumber) {
+      buildNumber = currentBuildNumber;
+      status = statusNew; // Reset status
```
Comment (keyonghan) [inline app_dart/lib/src/model/appengine/task.dart:278]: Shall this be the status of the current `build`?

File: app_dart/lib/src/model/appengine/task.dart [hunk 4]
```diff
@@ -262,27 +262,77 @@ class Task extends Model<int> {
...
+    createTimestamp = build.createdTimestamp?.millisecondsSinceEpoch;
+    startTimestamp = build.startedTimestamp?.millisecondsSinceEpoch;
+    endTimestamp = build.completedTimestamp?.millisecondsSinceEpoch;
+
+    _setStatusFromLuciStatus(build);
+  }
+
+  /// Get a [Task] status from a LUCI [Build] status/result.
+  String _setStatusFromLuciStatus(Build build) {
+    // Updates can come out of order. Ensure completed statsues are kept.
```
Comment (keyonghan) [inline app_dart/lib/src/model/appengine/task.dart:304]: Nit: statuses

Comment (keyonghan) [inline app_dart/lib/src/model/appengine/task.dart:304]: Nit: statuses

File: app_dart/test/model/task_test.dart [hunk 5]
```diff
@@ -55,6 +56,100 @@ void main() {
       );
       expect(Task.chromebot(commitKey: commitKey, createTimestamp: 123, builder: builder).isFlaky, isFalse);
     });
+
+    group('updateFromBuildPushMessage', () {
```
Comment (keyonghan) [inline app_dart/test/model/task_test.dart:60]: updateFromBuild

File: app_dart/lib/src/model/appengine/task.dart (hunk 2 above)
Comment (CaseyHillers) [reply app_dart/lib/src/model/appengine/task.dart]: Done

File: app_dart/lib/src/model/appengine/task.dart (hunk 4 above)
Comment (CaseyHillers) [reply app_dart/lib/src/model/appengine/task.dart]: Thanks, fixed

File: app_dart/test/model/task_test.dart (hunk 5 above)
Comment (CaseyHillers) [reply app_dart/test/model/task_test.dart]: Done

File: app_dart/lib/src/model/appengine/task.dart (hunk 1 above)
Comment (CaseyHillers) [reply app_dart/lib/src/model/appengine/task.dart:268]: No, it's seems to be specific to the push messages (I was able to verify this from the GCP logs)

**Review by keyonghan — APPROVED** [review]
LGTM

**Pull request merged**